plugins/suggested-blocks/src/index.js

```diff
@@ -0,0 +1,204 @@
+    }
+    uniqueRecentBlocks.sort((a, b) => recencyMap[a] - recencyMap[b]);
+    return this.generateBlockData(uniqueRecentBlocks);
+  }
+
+  generateBlockData = function(blockTypeList) {
+    const blockList = [];
+    for (const key of blockTypeList.slice(0, NUM_BLOCKS_PER_CATEGORY)) {
+      const json = (this.defaultJsonForBlockLookup[key] || {});
+      json['kind'] = 'BLOCK';
+      json['type'] = key;
+      json['x'] = null;
+      json['y'] = null;
+      blockList.push(json);
+    }
+    if (blockList.length == 0) {
+      blockList.push({
+        'kind': 'LABEL',
+        'text': 'No blocks have been used yet!',
+      });
+    }
+    return blockList;
+  }
+
+  /**
+   * Loads the state of this object from a serialized JSON.
+   * @param {object} data the serialized data payload to load from
+   */
+  loadFromSerializedData(data) {
+    this.defaultJsonForBlockLookup = data.defaultJsonForBlockLookup;
+    this.recentlyUsedBlocks = data.recentlyUsedBlocks;
+  }
+
+  /**
+   * Saves the state of this object to a serialized JSON.
+   * @return {object} a serialized data object including this object's state
+   */
+  saveToSerializedData() {
+    return {
+      defaultJsonForBlockLookup: this.defaultJsonForBlockLookup,
+      recentlyUsedBlocks: this.recentlyUsedBlocks,
+    };
+  }
+
+  /**
+   * Resets the internal state of this object.
+   */
+  clearPriorBlockData() {
+    this.defaultJsonForBlockLookup = {};
+    this.recentlyUsedBlocks = [];
+  }
+
+  /**
+   * Callback for when the workspace sends out events.
+   * @param {any} e the event object
+   */
+  eventListener(e) {
+    if (e.type == Blockly.Events.BLOCK_CREATE &&
+      this.workspaceHasFinishedLoading) {
+      const newBlockType = e.json.type;
+      // If this is the first time creating this block, store its default
+      // configuration so we know how exactly to render it in the toolbox
+      if (!this.defaultJsonForBlockLookup[newBlockType]) {
+        this.defaultJsonForBlockLookup[newBlockType] = e.json;
+      }
+      this.recentlyUsedBlocks.unshift(newBlockType);
+    } else if (e.type == Blockly.Events.FINISHED_LOADING) {
+      this.workspaceHasFinishedLoading = true;
+    }
+  }
+}
+
+export const init = function(workspace) {
+  const suggestor = new BlockSuggestor();
+  workspace.registerToolboxCategoryCallback('MOST_USED', suggestor.getMostUsed);
+  workspace.registerToolboxCategoryCallback('RECENTLY_USED',
+      suggestor.getRecentlyUsed);
+  workspace.addChangeListener(suggestor.eventListener);
+  suggestorLookup.set(workspace, suggestor);
+};
+
+/**
+ *
+ */
+class BlockSuggestorSerializer {
+  /** Constructs the block suggestor serializer */
+  constructor() {
+    /**
+     * The priority for deserializing block suggestion data.
+     * Should be less than the priority for blocks so that this state is
+     * applied after the blocks are loaded.
+     * @type {number}
+     */
+    this.priority = Blockly.serialization.priorities.BLOCKS - 10;
```

Comment:
People using an API the way I intended :DDD

Reply:
yay :)